### ISSUE TITLE
server side rendering className mismatch

### ISSUE THREAD

**Issue by L42y** [smyte/jsxstyle#80]
Not sure if I should put question at here, if not appropriate, please let me know.

I'm trying to use `react-router` v4 along with `react-apollo` to do server side rendering, following is the client side bootstrapping code:

```javascript
const Analysed = withGA(App);

render((
  <ApolloProvider
    client={client}>
    <BrowserRouter>
      <Analysed/>
    </BrowserRouter>
  </ApolloProvider>
), document.getElementById('root'));
```

and server side rendering code:

```javascript
        const ctx = {};
        const app = (
          <ApolloProvider
            client={client}>
            <StaticRouter
              context={ctx}
              location={request.url.href}>
              <App/>
            </StaticRouter>
          </ApolloProvider>
        );
        const markup = renderToString(app);

        if (ctx.url) {
          reply.redirect(ctx.url);
        } else {
          getDataFromTree(app).then(() => {
            const initialState = {
              apollo: client.getInitialState()
            };

            const stringifiedState =
                  JSON.stringify(initialState).replace(/</g, '\\u003c');
            reply(template.replace('<div id="root"></div>', renderToStaticMarkup(
              <div>
                <div
                  id="root"
                  dangerouslySetInnerHTML={{
                    __html: markup
                  }}/>

                <script
                  dangerouslySetInnerHTML={{
                    __html: `window.__APOLLO_STATE__ = ${stringifiedState};`
                  }}/>
              </div>
            )));
          });
        }
```

and simplified `<App>` component code:
```javascript
import {Col} from 'jsxstyle';

export default class App extends Component {
  render() {
    return (
      <Col>
        ...
        <Switch>
           <Route
             .../>
        </Switch>
      </Col>
    )
  }
}
```

Page rendered on client side with following error:

```
Warning: React attempted to reuse markup in a container but the checksum was invalid. This generally means that you are using server rendering and the markup generated on the server was not what the client was expecting. React injected new markup to compensate which works but you have lost many of the benefits of server rendering. Instead, figure out why the markup being generated is different on the client or server:
 (client) <div class="_x5" data-reactroot=""
 (server) <div class="_x2" data-reactroot=""
```

Just wondering what could be the cause makes client and server's class different

**Comment by meyer** [comment]
@L42y hmm, this is strange. My guess is that components are being iterated over in a non-deterministic order. The class name suffix is generated with an incrementing integer and is not content-based. As a workaround, I can switch `jsxstyle-loader` to use content-based class names, just like we do with runtime `jsxstyle`. Class names will be longer but will not change between renders.

**Comment by L42y** [comment]
@meyer maybe a config to switch between incrementing integer and content-based suffix?

**Comment by meyer** [comment]
@L42y sounds good to me.

**Comment by meyer** [comment]
@L42y I just added deterministic class name support to `jsxstyle-loader` in 8caa55399b9a1a71bc387a4aa1eb88f34e6b05f2. I’ll publish a new version of `jsxstyle-loader` as soon as I’m satisfied with my local testing.

To enable the option: add `deterministic: true` to your loader config object. If you’ve got a better name for the option let me know. I’m not 100% sold on `deterministic`.

**Comment by L42y** [comment]
@meyer thanks a lot for your great work! I just tried the new version, no more class prop mismatch.

about the option naming, I don't have a better name neither, but an option with string value like `content-hash` instead of boolean also ok for me, maybe this would make the option naming easier?

**Comment by meyer** [comment]
@L42y yeah, I had thought about something like `classNameFormat`, where people could specify a template (kind of like [webpack’s substitutions][webpack]). The “template” right now would be something like `_x[contenthash]` for deterministic rendering and `_x[base36]` incrementing integer-based class names.

[webpack]: https://webpack.js.org/configuration/output/#output-filename

**Comment by meyer** [comment]
I think for simplicity’s sake I’m going to go with `classNameFormat: 'hash'` as the content hash option and default it to `'base36'`. Template stuff can come later if necessary.

**Comment by meyer** [comment]
@L42y I’ve added and documented the `classNameFormat` setting and just published the changes as `jsxstyle-loader@1.0.4`. Thanks for your input! 🎉🎉🎉

**Comment by L42y** [comment]
@meyer thanks for your great work!
